runtime/src/bank.rs

```diff
@@ -5176,7 +5176,17 @@ impl Bank {
                 .iter()
                 .chain(additional_builtins.unwrap_or(&[]).iter())
             {
-                if builtin.enable_feature_id.is_none() {
+                // The builtin should be added if it has no enable feature ID
+                // and it has not been migrated to Core BPF.
+                //
+                // If a program was previously migrated to Core BPF, accountsDB
+                // from snapshot should contain the BPF program accounts.
+                let builtin_is_bpf = |program_id: &Pubkey| {
+                    self.get_account(program_id)
+                        .map(|a| a.owner() == &bpf_loader_upgradeable::id())
+                        .unwrap_or(false)
+                };
+                if builtin.enable_feature_id.is_none() && !builtin_is_bpf(&builtin.program_id) {
```

Comment:
Unfortunately this feature activation isn't idempotent.

Since the migration path is fallible, we can't say for certain the actual action triggered by the feature activation was successful. If we were to key on feature activation here, we could end up _not_ adding a builtin when no BPF counterpart exists, resulting in no program at the address at all.

Reply:
Or we do it transactional and if it fails we never update the program accounts, effectively ignoring the feature activation.

Reply:
How are you thinking this might work? Send a transaction via CLI similar to feature activation?

Reply:
Delay all account updates until the very last verification step, once no more errors can be thrown.

Reply:
> Delay all account updates until the very last verification step, once no more errors can be thrown.

Yep, this is precisely how I've implemented the migration bit. If any fallible step of the migration fails, an error is returned, and that's handled gracefully by leaving `builtin_is_bpf` as `false`.
https://github.com/anza-xyz/agave/pull/590/files#diff-ed47b4a0198313377e091bb3957bbbc63d937805426d1b2b6de39d0a50d32a0cR6576-R6577

However, here in `finish_init`, we need to determine whether or not the program is BPF, and just checking if the feature to migrate it was activated isn't sufficient, because that migration could have failed on a previous epoch boundary.

Reply:
A migration only fails when the setup for the source program was not done properly. In that case a rekeying and a new feature gate should be done, instead of retrying the same in the next epoch being uncontrolled.

Reply:
I've designed this so that only `new_feature_activations` will trigger a migration, to serve that point precisely. In the case of a failed migration, the feature itself still becomes active (in the sense of the feature account), but the migration never occurs.

With the runtime checks I've put in place here, bank views this as completely acceptable, and doesn't attempt another migration next epoch because the feature is no longer "new".

Then a contributor must, as you said, re-key the feature and place the new one on the builtin's migration config.